pkg/rebuild/rebuild/rebuildremote.go

```diff
@@ -91,6 +93,110 @@ var rebuildContainerTpl = template.Must(
 				`)[1:], // remove leading newline
 	))
 
+var standardBuildTpl = template.Must(
+	template.New(
+		"standard build",
+	).Parse(
+		textwrap.Dedent(`
+				set -eux
+				cat <<'EOS' | docker buildx build --tag=img -
+				{{.Dockerfile}}
+				EOS
+				docker run --name=container img
+				`)[1:], // remove leading newline
+	))
+
+// NOTE(impl): There are a number of factors complicating this harness that warrant some explanation.
+//   - Overview: The proxy and build are executed in separate containers. All
+//     HTTP traffic is redirected to the proxy from the build AND containers
+//     created from the build (i.e. buildx workers).
+//   - Network basics: The top-level container uses the pre-allocated
+//     "cloudbuild" network on GCB. We create a separate "proxynet" network to
+//     bridge the build and the proxy. The proxy is also connected to the
+//     "cloudbuild" network so its admin server is accessible from the
+//     top-level container to retrieve the network log.
+//   - Network namespaces: We use iptables to redirect traffic to the proxy.
+//     iptables is configured per-network namespace. Crucially, Docker networks
+//     are not the same as network namespaces! By default, though, each
+//     container is allocated its own namespace even if they're connected to
+//     the same network. So to apply iptables rules to the build container, we
+//     need to ensure they all execute the iptables rules (which requires root)
+//     in that same container. But to ensure the same for all derivative
+//     containers of the build, we need those same iptables rules. Luckily,
+//     Docker does allow you to specify "container:<NAME>" as the container
+//     network which joins both the same network AND the same network
+//     namespace. So we use the proxy to enforce that all new containers use
+//     the "container:build" network.
+//   - User setup: We use the iptables' owner module feature of redirecting
+//     traffic based on the UID of the originating process. To do so here,
+//     though, we need a user/uid that's shared across the proxy and build
+//     containers. The user namespaces are not shared between the containers so
+//     we cannot refer to the user by name from one if it's created by the
+//     other. Creating a user at the top-level, recreating it in the proxy, and
+//     using the uid in the build satisfies our constraints.
```

Comment:
Are you worried at all that someone may create a user inside the container that isn't caught by the iptables rules?

Reply:
So this rule means all users except the proxy get redirected to the proxy so I think the only viable attack would be if the build was able to escalate privilege and execute code as the proxy user. That's definitely a concern but many other parts of this security model break down at that point.

Reply:
Ah sorry, I'd missed that, that addresses my concern!